DistroLauncher/ExitStatus.cpp

```diff
@@ -0,0 +1,203 @@
+            return;
+        }
+        prefixedFilePath.append(distroName.value());
+        prefixedFilePath.append(launcherCommandFilePath);
+        std::ifstream launcherCmdFile;
+        launcherCmdFile.open(prefixedFilePath);
+        if (launcherCmdFile.fail()) {
+            if (errno != ENOENT) {
+                Helpers::PrintErrorMessage(HRESULT_FROM_WIN32(GetLastError()));
+            }
+            // File not found will be interpreted as there is nothing to do.
+            return;
+        }
+
+        auto parserResult = internal::parseExitStatusFile(launcherCmdFile);
+        if (parserResult.has_value()) {
+            // Finally take the actions and delete the file.
+            const KeyValuePairs& launcherCmds{parserResult.value()};
+            auto actionResult = act(launcherCmds);
+            auto configResult = config(launcherCmds);
+        } else {
+            wprintf(parserResult.error());
+        }
+
+        launcherCmdFile.close();
+        // We don't want that file existing after actions were taken.
+        std::filesystem::remove(prefixedFilePath);
+
+    }
+
+    namespace
+    {
+        bool ensureDistroStopped(unsigned int maxNoOfRetries);
+
+        KeyValuePairs launcherCmds;
+
+        // Shall we need to extend the capabilities triggered by the launcher-command file, here is what we need to do:
+        // 1. Add more functions inside the Actions namespace with the same signature of the existing ones.
+        // 2. Add those functions associated with the expected value in the file which will trigger them to the
+        // capabilities constant map.
+        // 3. If that requires extending the parser grammar, see `exit_status_parser.cpp`
+        namespace Actions
+        {
+            VoidResult rebootDistro();
+            VoidResult shutdownDistro();
+        };
+        using Action = VoidResult (*)();
+        const std::unordered_map<std::string_view, Action> capabilities{{"reboot", &Actions::rebootDistro},
+                                                                        {"shutdown", &Actions::shutdownDistro}};
+
+        // Polls WSL to ensure the distro is actually stopped.
+        bool ensureDistroStopped(unsigned int maxNoOfRetries)
+        {
+            for (unsigned int i = 0; i < maxNoOfRetries; ++i) {
+                auto runner = Helpers::ProcessRunner(L"wsl -l --quiet --running");
+                auto exitCode = runner.run();
+                if (exitCode != 0L) {
+                    return false;
+                }
+
+                // Returns true once we don't find the DistributionInfo::Name in process's stdout
+                auto output = runner.getStdOut();
+                if (output.find(DistributionInfo::Name) == std::wstring::npos) {
+                    return true;
+                }
+
+                // We don't need to be hard real time precise.
+                Sleep(997); // NOLINT(readability-magic-numbers): only used here.
```

Comment:
why exactly 997? :D

Reply:
Just o be fancy and not promissing exactly one second delay in between checks. 😄